validator/src/scheduler/mod.rs

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use proto::batch::Batch;
+use proto::events::Event;
+use proto::transaction_receipt::StateChange;
+
+pub mod py_scheduler;
+
+pub trait Scheduler {
+    /// Add a batch to the scheduler, optionally specifying that the transactions
+    /// in this batch and each of the batches in order up to this one should produce a
+    /// Merkle root specified by expected_state_hash. If the two roots are equal, the results of
+    /// the block are written to the database, otherwise not.
+    fn add_batch(
+        &mut self,
+        batch: Batch,
+        expected_state_hash: Option<&str>,
+    ) -> Result<(), SchedulerError>;
+
+    /// Signal to the scheduler that it can finish it's work and no
+    /// more batches will be handed to it.
+    fn finalize(&mut self, unschedule_incomplete: bool) -> Result<(), SchedulerError>;
+
+    /// Cancel the scheduling of the supplied transactions as they are no longer needed.
+    fn cancel(&mut self) -> Result<(), SchedulerError>;
+
+    /// Ask if the ExecutionResults are ready, optionally blocking until they become available.
+    fn complete(&mut self, block: bool) -> Result<Option<ExecutionResults>, SchedulerError>;
+}
+
+pub struct ExecutionResults {
+    pub beginning_state_hash: Option<String>,
+    pub ending_state_hash: Option<String>,
+    pub batch_results: Vec<(String, Vec<TxnExecutionResult>)>,
+}
+
+pub struct TxnExecutionResult {
+    pub signature: String,
+    pub is_valid: bool,
+    pub state_changes: Vec<StateChange>,
+    pub events: Vec<Event>,
+    pub data: Vec<(String, Vec<u8>)>,
+    pub error_message: String,
+    pub error_data: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub enum SchedulerError {
+    /// The scheduler transition is not allowed by the Finite State Machine.
+    FSMError(String),
+    /// There was an error caused by the FFI interface.
+    FFIError(String),
+    /// There was a mismatch that wasn't a Python error, but the Rust should
+    /// treat as an error because, for example, a value was None when it shouldn't be,
+    /// or information was missing.
+    PythonRustMismatch(String),
```

Comment:
FFIError and PythonRustMismatch are not good error types for Scheduler as they leak details about how a particular implementation of this trait works.

If there are things that are technically possible based on the library we are using but should never happen in practice, it is probably okay to just panic on them instead of propagating the error. I suspect these errors may go away if you use that pattern. For example, if we create and hold onto a python object in the Rust constructor, that object's type should never change. So if later we go to call a method on that object and that method does not exist, we should probably just panic at that point because either some memory has been corrupted or there is a serious programming error (like the python code changed and we forgot to update it in the Rust code).

Reply:
I agree that some of the python errors -> Rust errors should just fail, but I'm sure there are python errors that should be propagated up that wouldn't normally be part of the Rust Api.

Reply:
Right, but we need an error type that makes sense for these that isn't implementation specific.